corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views/query.js

```diff
@@ -417,18 +421,22 @@ hqDefine("cloudcare/js/formplayer/menus/views/query", function () {
         submitAction: function (e) {
             e.preventDefault();
             FormplayerFrontend.trigger('clearNotifications', errorHTML, true);
-
             var invalidFields = [];
             this.children.each(function (childView) {
                 if (!childView.isValid()) {
                     invalidFields.push(childView.model.get('text'));
+                    FormplayerFrontend.trigger('showInputError', childView.model.get("error") ||
+                    "Please enter a value for this field.", childView.model.get("text"));
```

Comment:
It will show "Please enter a value for this field." for required field that are blank and whatever the custom error message is for other invalid inputs.

Reply:
Hmm, that's a good point - I guess we don't really have a way to distinguish anyways.  I made the configuration such that messages are not required for validation or required conditions, figuring we'd use a default for each.  @shubham1g5 what do you think?

Reply:
@esoergel Good that we are discussing this since it looks like there is some gap between FP and HQ implementations on this. 

I have treated the validation message as a [required element](https://github.com/dimagi/commcare-core/blob/queryPromptRequired/src/main/java/org/commcare/xml/QueryPromptParser.java#L119) while kept the required message optional. My reasoning being it's quite hard to give a helpful default validation error in comparison to required error that almost always says 'this field is empty when it should not be'. Although I looked at how Xform works and it seems like we do have a default validation message there which just says 'This answer is outside allowed range'. So probably best to do the same here ?

Both validations and required prompts are communicated differently via FP APIs so it should be possible to separate the two out.

Reply:
I have made the validation message optional on Formplayer side as well(https://github.com/dimagi/commcare-core/pull/1138/commits/60d95d35eb5f004f6560aa73d361e57bdb7e3f8f) and I am now sending a [default message ](https://github.com/dimagi/commcare-core/pull/1138/commits/6f5af7f0292b83927de06431c24ce61bf236fc37) into `errors` when there is no message defined.

Reply:
Great - I also like the idea of using xforms/vellum as a reference for this sort of question, as that's a clear parallel.  (relatedly, if vellum doesn't support multiple validation conditions for a single question, why should case search?  Oh well)

So does this mean that formplayer will always provide a message here, then?

Reply:
> So does this mean that formplayer will always provide a message here, then?

that's right.